### PR TITLE
feat(auth): fetchDevices API

### PR DESCRIPTION
**Pull request by Samaritan1011001**
#### Description of changes

Implements fetchDevices API for V6. 

Relevant Cognito API ref: https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ListDevices.html

#### Description of how you validated changes
1. Enabled always remember option for the user pool
2. Logged in a user via V5 UI authenticator
3. Used the implemented V6 fetchDevices API
4. Validated it returned the devices that were remembered
5. Further checks via unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/__tests__/providers/cognito/fetchDevices.test.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	});
+	test('fetchDevices API should return a valid deviceId and deviceName', async () => {
+		const result = await fetchDevices();
+		expect(result).toEqual(
+			deviceManagementAPIConstants.ListDevicesAPIResultHappy
+		);
+	});
+});
+
+describe('fetchDevices API Edge Cases:', () => {
```
Comment (israx) [inline packages/auth/__tests__/providers/cognito/fetchDevices.test.ts:45]: Is this guaranteed ? What are the circumstances for the service to sometimes return a name and sometimes not ? Is this edge case acknowledged by cognito or seen in a github issue ? If not, I don't think is worth assuming it.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+const MAX_DEVICES = 60;
+
+/**
+ * Fetch devices associated with an authenticated user
+ *
+ * @returns AuthDevice[]
+ * @throws validation: {@link AuthError}
+ *
+ */
+export async function fetchDevices(): Promise<AuthDevice[]> {
```
Comment (israx) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:25]: as discussed async, this should be `AWSAuthDevice`

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	const session = await fetchAuthSession({});
+	if (!session.tokens) {
```
Comment (israx) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:27]: This could be a pretty common error for APIs that require an AccessToken.

I think we should standardize and be consistent how we throw this validation error. 

Currently some APIs and token provider have been using the `assertAuthTokens` util. That's one source of truth to handle missing tokens instead of having different error messages and codes for each API.

Probably we can just add a recovery suggestion in that util that is generic for all APIs.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		});
+	}
+	const requestParams: ListDevicesCommandInput = {
+		AccessToken: session.tokens.accessToken.toString(),
+		Limit: MAX_DEVICES,
+	};
+
+	const authConfig = AmplifyV6.getConfig().Auth;
+	const userPoolId = authConfig?.userPoolId;
+	if (!userPoolId) {
```
Comment (israx) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:42]: The same claim about being consistent how we throw config errors across all APIs. 

In addition, I think we should check for this error first before checking if there are tokens first.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		throw new AuthError({
+			name: 'AuthConfigException',
+			message: 'Cannot list devices without an userPoolId',
+			recoverySuggestion:
+				'Make sure a valid userPoolId is given in the config.',
+		});
+	}
+
+	const region = userPoolId.split('_')[0];
```
Comment (israx) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:51]: the other APIs are using the `getRegion` util, which throws if the region has an inconsistent format. We could reuse it here.

File: packages/auth/src/types/Auth.ts [hunk 6]
```diff
@@ -215,6 +215,11 @@ export interface IAuthDevice {
 	name: string;
 }
 
+// TODO(V6): Decide which type to keep, this or the IAuthDevice
+export type AuthDevice = {
```
Comment (israx) [inline packages/auth/src/types/Auth.ts:219]: I think the `types` on this file are mostly v5.  The `AuthDevice` should be on `auth/types/models.ts` and the upcoming `AWSAuthDevice` on `auth/cognito/types/models`

**Review by israx — COMMENTED** [review]
I just have some general comments:

1. we should standardize and be consistent on config and token errors for all APIs
2. The location of the `AuthDevice` type.

File: packages/auth/__tests__/providers/cognito/fetchDevices.test.ts (hunk 1 above)
Comment (Samaritan1011001) [reply packages/auth/__tests__/providers/cognito/fetchDevices.test.ts:45]: Service call won't return a name if we did not pass in a `deviceName` via `confirmDevice` API in the first place during signIn. So until we make sure to pass that in, this field will always be present. I added this edge case handling because if there is no attribute with the Name as `device_name` in the service call response we may return empty string or not return the field at all.  But I see other platforms have it non-optional so will align and update this.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+// Cognito Documentation for max device
+// tslint:disable-next-line:max-line-length
+// https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ListDevices.html#API_ListDevices_RequestSyntax
+const MAX_DEVICES = 60;
+
+/**
+ * Fetch devices associated with an authenticated user
+ *
+ * @returns AuthDevice[]
+ * @throws validation: {@link AuthError}
```
[the remark this answers is not in the document]
Comment (israx) [reply packages/auth/src/providers/cognito/apis/fetchDevices.ts:24]: ditto

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+			}
+			if (device.DeviceCreateDate)
+				deviceInfo.createDate = device.DeviceCreateDate;
+			if (device.DeviceLastAuthenticatedDate)
+				deviceInfo.lastAuthenticatedDate = device.DeviceLastAuthenticatedDate;
+			if (device.DeviceLastModifiedDate)
+				deviceInfo.lastModifiedDate = device.DeviceLastModifiedDate;
+			return deviceInfo;
+		} else {
+			throw new AuthError({
```
Comment (israx) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:67]: I wonder if it make sense just to return an empty object instead `{}` ?

**Review by israx — COMMENTED** [review]
Just a comment about improving the actual doc  string on this API and about whether we should considered returning an empty object instead of throwing.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts (hunk 8 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/fetchDevices.ts:67]: I wonder if there might be the case where some elements on the device list might include `attributes` and `id` and in others not.

Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/apis/fetchDevices.ts:71]: Not sure we want to have empty objects in the array, maybe we can ignore the ones that do not have the deviceKey

File: packages/auth/__tests__/providers/cognito/testUtils/authApiTestParams.ts [hunk 9]
```diff
@@ -157,7 +178,8 @@ export const authAPITestParams = {
 	},
 	GuestIdentityId: { id: 'guest-identity-id', type: 'guest' },
 	PrimaryIdentityId: { id: 'primary-identity-id', type: 'primary' },
-
+	validAuthConfig: { ...validAuthConfig },
+	mandatorySignInEnabledConfig: { ...mandatorySignInEnabledConfig },
```
Comment (stocaaro) [inline packages/auth/__tests__/providers/cognito/testUtils/authApiTestParams.ts:182]: What are we getting from the spread operator here as opposed to just assigning the object directly?  Probably a reason, but its not intuitive to me looking at it.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+			};
+			if (!!deviceName) {
+				deviceInfo.deviceName = deviceName;
+			}
+			if (device.DeviceCreateDate)
+				deviceInfo.createDate = device.DeviceCreateDate;
+			if (device.DeviceLastAuthenticatedDate)
+				deviceInfo.lastAuthenticatedDate = device.DeviceLastAuthenticatedDate;
+			if (device.DeviceLastModifiedDate)
+				deviceInfo.lastModifiedDate = device.DeviceLastModifiedDate;
```
Comment (stocaaro) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:68]: nit: Can we use null-coalescant or optional assignment instead of conditioning each line? I guess reading these it's a bit more  complicated as we're conditioning on the right hand side as opposed to the left, but one lining these as something like the following seems more efficient / concise:

```ts
deviceInfo.createDate = device.DeviceCreateDate ?? deviceInfo.createDate;
```

File: packages/auth/__tests__/providers/cognito/credentialsProvider.test.ts [hunk 11]
```diff
@@ -105,7 +84,9 @@ describe('Guest Credentials', () => {
 					return authAPITestParams.NoAccessKeyCredentialsForIdentityIdResult as cogId.GetCredentialsForIdentityOutput;
 				}
 			);
-			configSpy.mockImplementationOnce(() => mandatorySignInEnabledConfig);
+			configSpy.mockImplementationOnce(
+				() => authAPITestParams.mandatorySignInEnabledConfig
+			);
```
Comment (stocaaro) [inline packages/auth/__tests__/providers/cognito/credentialsProvider.test.ts:89]: Seems odd that the test files are space indented where the src files appear to be tab indented. Is our linter borked?

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	const authConfig = AmplifyV6.getConfig().Auth;
+	const region = getRegion(authConfig?.userPoolId);
+
+	const res = await listDevices({ region }, requestParams);
+	const devices: DeviceType[] = res.Devices ?? [];
+
+	return devices.map(device => {
+		if (device.DeviceKey && device.DeviceAttributes) {
+			let deviceName: string | undefined;
+			const deviceInfo: AWSAuthDevice = {
```
Comment (stocaaro) [inline packages/auth/src/providers/cognito/apis/fetchDevices.ts:49]: This map is pretty chunky and hard to parse its intent. I would think about adding a private function to this file that performs the transformation on device, where the function name serves to make clear the transformation.

**Review by stocaaro — COMMENTED** [review]
Seems close. I've made some comments, mostly nits.

File: packages/auth/src/providers/cognito/apis/fetchDevices.ts (hunk 10 above)
Comment (AllanZhengYP) [reply packages/auth/src/providers/cognito/apis/fetchDevices.ts:68]: @stocaaro I think @Samaritan1011001 is trying to prevent the key being set to the `deviceInfo` object if the SDK output does not exist. Instead of if statement, you can use this:
```ts
const deviceInfo: AWSAuthDevice = {
  //..
  ...(device.DeviceCreateDate ? { createDate: device.DeviceCreateDate } : {}),
}
```

**Comment by stocaaro** [comment]
This needs to be recreated or redirected to `next/main`.
